string_grouper/string_grouper.py

```diff
@@ -256,34 +274,224 @@ def n_grams(self, string: str) -> List[str]:
+        max_n_matches = self._max_n_matches
+
+        self._true_max_n_matches = 0
+        for left_block in block_ranges_left:
+            left_matrix = self._get_left_tf_idf_matrix(left_block)
+            nnz_rows = np.full(left_matrix.shape[0], 0, dtype=np.int32)
+            for right_block in block_ranges_right:
+                self._max_n_matches = min(
+                    right_block[1] - right_block[0],
+                    max_n_matches
+                )
+                right_matrix = self._get_right_tf_idf_matrix(right_block)
+
+                # Calculate the matches using the cosine similarity
+                matches, block_true_max_n_matches = self._build_matches(
+                    left_matrix, right_matrix, nnz_rows
+                )
+
+                # build match-lists from matrix
+                r, c = matches.nonzero()
+                d = matches.data
+                (self._r, self._c, self._d) = (
+                    np.append(self._r, r + left_block[0]),
+                    np.append(self._c, c + right_block[0]),
+                    np.append(self._d, d)
+                )
+                # end of inner loop
+
+            self._true_max_n_matches = \
+                max(block_true_max_n_matches, self._true_max_n_matches)
+            # end of outer loop
+
+        self._max_n_matches = max_n_matches
+        return max(n_blocks) > 1
+
+    def _fit_blockwise_auto(self,
+                            left_partition=(None, None),
+                            right_partition=(None, None),
+                            nnz_rows=None,
+                            whoami=0):
+        # This is a recursive function!
+        # fit() has been extended here to enable StringGrouper to handle large
+        # datasets which otherwise would lead to an OverflowError
+        # The handling is achieved using block matrix multiplication.
+        def begin(partition):
+            return partition[0] if partition[0] is not None else 0
+
+        def end(partition, left=True):
+            if partition[1] is not None:
+                return partition[1]
+
+            return len(self._left_DataFrame if left else self._right_DataFrame)
+
+        left_matrix = self._get_left_tf_idf_matrix(left_partition)
+        right_matrix = self._get_right_tf_idf_matrix(right_partition)
+
+        if whoami == 0:
+            # At the topmost level of recursion initialize nnz_rows
+            # which will be used to compute _true_max_n_matches
+            nnz_rows = np.full(left_matrix.shape[0], 0, dtype=np.int32)
+            self._true_max_n_matches = 0
+
+        try:
+            # Calculate the matches using the cosine similarity
+            matches, true_max_n_matches = self._build_matches(
+                left_matrix, right_matrix, nnz_rows[slice(*left_partition)])
+        except OverflowError:
+            # Matrices too big!  Try splitting:
+            left_matrix = None
```

Comment:
left_matrix / right_matrix variables are not getting used

Reply:
@Bergvca 

Indeed, in the Exception handler, left_matrix and right_matrix are no longer being used.  And since this is a recursive function, I didn’t want them residing in memory hogging the memory space while the function continues to recurse which would lead to degradation in performance.  So I set them to None.  

(Every new call to fit_blockwise_auto() recreates it own copy of these variables and they remain in memory until the end of the recursion.)

Reply:
Ah I understand now. To make it more clear you could also do `del left_matrix`